Sources/ResilientDecoding/ErrorReporting.swift

```diff
@@ -1,43 +1,114 @@
+    self[resilientDecodingErrorReporterCodingUserInfoKey] = errorReporter
+    return errorReporter
+  }
+
+}
+
+extension JSONDecoder {
+
+  /**
+   Creates and registers a `ResilientDecodingErrorReporter` with this `JSONDecoder`. Any `Resilient` properties which this `JSONDecoder` decodes will report their errors to the returned error reporter.
+   - note: May only be called once per `JSONDecoder`
+   */
+  public func enableResilientDecodingErrorReporting() -> ResilientDecodingErrorReporter {
+    userInfo.enableResilientDecodingErrorReporting()
+  }
+
+}
+
+extension Decoder {
+
+  /**
+   This method should be called whenever an error is handled by the `Resilient` infrastructure.
+   Care should be taken that this is called on the most relevant `Decoder` object, since this method uses the `Decoder`'s `codingPath` to place the error in the correct location in the tree.
+   */
+  func resilientDecodingHandled(_ error: Swift.Error) {
+    guard let errorReporterAny = userInfo[resilientDecodingErrorReporterCodingUserInfoKey] else {
+      return
+    }
+    /**
+     Check that we haven't hit the very unlikely case where someone has overriden our user info key with something we do not expect.
+     */
+    guard let errorReporter = errorReporterAny as? ResilientDecodingErrorReporter else {
+      assertionFailure()
+      return
+    }
+    errorReporter.resilientDecodingHandled(error, at: codingPath.map { $0.stringValue })
+  }
+
+}
+
+// MARK: - Accessing Reported Errors
 
 public final class ResilientDecodingErrorReporter {
 
   /**
    This is meant to be called immediately after decoding a `Decodable` type from a `Decoder`.
    - returns: Any errors encountered up to this point in time
    */
-  public func flushReportedErrors(includeUnknownNovelValueErrors: Bool = false) -> [Error] {
-    let reportedErrors = errorContainer.errors
-    errorContainer = ErrorContainer()
-    let errors: [Error]
-    if isMissingReportedErrors {
-      errors = [IsMissingReportedErrors()] + reportedErrors
-    } else {
-      errors = reportedErrors
-    }
-    return errors.filter { includeUnknownNovelValueErrors || !($0 is UnknownNovelValueError) }
+  public func flushReportedErrors() -> ErrorDigest? {
+    let digest = hasErrors ? currentDigest : nil
+    hasErrors = false
+    currentDigest = ErrorDigest()
+    return digest
   }
 
   /**
    This should only ever be called by `Decoder.resilientDecodingHandled` when an error is handled, consider calling that method instead.
    It is `internal` and not `fileprivate` only to allow us to split the up the two files.
    */
   func resilientDecodingHandled(_ error: Error, at path: [String]) {
-    errorContainer.insert(error, at: path)
+    hasErrors = true
+    currentDigest.root.insert(error, at: path)
+  }
+
+  fileprivate var currentDigest = ErrorDigest()
+  private var hasErrors = false
+
+}
+
+public struct ErrorDigest {
+
+  public var errors: [Error] { errors(includeUnknownNovelValueErrors: false) }
```

Comment:
If we only invoke `error(includeUnknownNovelValueErrors:)` once and only pass `false`, why have the `includeUnknownNovelValueErrors` parameter at all?

Reply:
Nvmd `error(includeUnknownNovelValueErrors:)` is public 👍